enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,891 @@
+
+#### Story 3
+
+As a cluster administrator, I would like to test Cluster API before rolling it out to my production clusters, so that I can build confidence in the new APIs.
+To do this, I would like to be able to add Cluster API MachineSets to add new Machines to my Cluster without having to have an equivalent Machine API MachineSet to match it.
+
+#### Story 4
+
+As a developer of Machine related APIs, I would like to be able to gather feedback on Machine conversion layers before forcing new APIs onto users.
+By gathering feedback, we can ensure the conversion is working, and that we are not breaking existing users when they are upgraded into newer versions
+of OpenShift that rely on Cluster API.
+
+### Goals
+
+* Allow users to migrate Machines between Machine API and Cluster API to test the new API
+* Allow users a choice (initially[^3]) of when they want to migrate
+* Eventually[^3] remove the Machine API controllers and leverage the migration layer to convert Machine API machines to Cluster API machines
+* Allow the Cluster Autoscaler, MachineHealthCheck and ControlPlaneMachineSet to continue to work during a semi-migrated cluster state
+* Ensure that every Machine API resource is translated to a Cluster API resource to allow users to test the new API
+* Allow access to newer features of Cluster API that do not exist in Machine API
+* Avoid carry patches in the Cluster API controllers to ensure that we can easily upgrade to newer versions of Cluster API
+
+[^3]: In the first iteration of this project, the goal is to allow users to migrate individual resources.
+In later iterations, the plan is to remove the Machine API controllers. The linked goals will be executed in separate releases.
+
+### Non-Goals
+
+* Setting up or configuring any part of Cluster API, this is handled in a separate enhancement
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new Machine API machines when new Cluster API MachineSets are being created[^4]
+* Port any feature from Cluster API into Machine API
+
+[^4]: We want to promote, over time, a reduction in the number of Machine API Machines, as such,
+any resources created directly in Cluster API, will not be reflected into Machine API, unless they have an owner,
+and, that owner is already reflected into Machine API.
+
+## Proposal
+
+We will implement a two-way sync controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using a new field on the Machine API resource, users will be able to choose which API is authoritative.
+Any discrepancy between the non-authoritative resource and the authoritative resource will be overwritten by the sync controller to match the authoritative resource.
+
+When the API is non-authoritative, the controllers should be paused (ignore the resource),
```

Comment:
This might be a nit, but I would remove the mention of paused entirely here, since the authoritativeness is per-resource, as I understand it. It's entirely possible, given this enhancement, that everything would be managed by MAPI (for example) except for 1 `Machine`. In that case, both MAPI and CAPI controllers are still running, it's just that the MAPI controllers are acting on more resources.

Reply:
Fair point, will reword this to make it more obvious that I'm referring to the API being authoritative for a single resource, and then the controllers should be paused for a particular resource

Reply:
I think I'm still hung up on the word `paused` :)

Stating that the resource is ignored is more clear to me. `Paused`, to me, implies that the controller isn't running. Perhaps this is a use of the term that I'm not fully familiar with; I do know CAPI refers to `pausing` clusters where it doesn't manage them.

Reply:
Coming back to this after a brief chat with Joel about this: `pause` is already used in Cluster API to mean "pausing management of a resource," which makes sense in this context now.

My reading was initially seeing "pausing a running machine." I think there is still room for misunderstanding here, but I also do not believe that its large enough to try to change semantics that already exist upstream, or to add our own.